### PR TITLE
feat(bug-report): let the reporter navigate before taking the screenshot

### PR DESCRIPTION
**Pull request by RyanGroch**
The form could only capture whatever was under the Help dialog. Asking for a screenshot now closes the dialog and leaves a bar across the bottom of the window, so the reporter can go to where the bug is, capture it from the bar, and come back to the form they left with the image on it.

## The bar

- Any opening of the dialog dismisses it, whether from the capture landing, Back to report, the sidebar's Help button or a crash report, so no path can leave it behind. Diagnostics are re-read when the form comes back, since the reporter may have reproduced the bug meanwhile.
- It publishes its height as a CSS variable. The main panel, the sidebar and its lists, the privacy banner and the toaster all subtract or offset by it, so nothing covers the bar and the bar covers nothing. The native preview composites above all DOM, so shrinking the layout is what keeps it visible.
- It sits above the app's z-50 layer, so it stays reachable over the fullscreen code view and the image lightbox.
- Capture takes keyboard focus when the bar appears, with the instructions described to it for screen readers. Escape goes back to the report while focus is inside the bar.
- The Help dialog now mounts in the root layout rather than the sidebar. React delivers a portal's pointer events to its React ancestors, so pointing at the bar was keeping a hover-expanded sidebar open.

## Paste reminders

The screenshot travels on the clipboard, and reports often arrive without it. When one was captured:

- The issue body gets a visible line under the Screenshot heading asking for the paste, where the reporter is looking when it has to happen.
- After filing, the dialog stays on a closing step asking whether the screenshot was pasted, with the preview and a Done button. It asks rather than asserts, because most reporters only read it once they are back from GitHub: it tells them how to paste, how to recover if they already submitted without it, and that they are done if it is already in.

A reminder after the issue exists needs GitHub's side, so it belongs in the triage workflow as a follow-up.

## Not supported

Content inside a modal dialog cannot be captured. A press on the bar lands outside the modal and closes it before the capture runs. A capture that is not a press, such as a keyboard shortcut, is the natural follow-up.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/dyad-sh/dyad/pull/4587?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/components/HelpDialog.tsx [hunk 1]
```diff
@@ -955,8 +986,22 @@ export function HelpDialog() {
...
+        onCancel={cancelCaptureBar}
+      />
+    ) : null;
+
   return (
     <>
+      {/* Out of the sidebar this dialog is mounted in: its container carries
+          a transform, which would make the bar's fixed position relative to
+          the sidebar rather than the window. */}
+      {captureBar && createPortal(captureBar, document.body)}
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/HelpDialog.tsx:1030]: Fixed in 1a10cc5c. The dialog now mounts in the root layout next to ForceCloseDialog instead of inside the sidebar, so the bar's pointer events no longer reach the sidebar's hover handlers through React's tree. The sidebar's Help button only sets the atom, so nothing else changed.

File: src/components/ScreenshotCaptureBar.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    };
+  }, []);
+
+  return (
+    <div
+      ref={root}
+      role="region"
+      aria-label={t("home:report.captureBarLabel")}
+      data-testid="screenshot-capture-bar"
+      className="fixed inset-x-0 bottom-0 z-20 flex items-center gap-3 border-t-2 border-primary/70 bg-(--background-lightest) px-4 py-2 shadow-[0_-2px_8px_rgba(0,0,0,0.08)]"
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Fixed in e38c5ec1. The privacy banner now offsets its bottom by the same --layout-bottom-bar-height variable the sidebar and main panel read, so it stacks above the bar instead of covering its buttons. Went with stacking rather than a higher z-index on the bar, since that would have hidden the banner's buttons instead.

File: src/components/ScreenshotCaptureBar.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      <p className="min-w-0 flex-1 text-sm">
+        <span className="font-medium">
+          {t("home:report.captureBarHeading")}
+        </span>{" "}
+        <span className="text-muted-foreground">
+          {t("home:report.captureBarHint")}
+        </span>
+      </p>
+      <div className="flex shrink-0 gap-1.5">
+        <Button variant="default" size="sm" onClick={onCapture}>
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Fixed in 249cd909. The Capture button takes focus when the bar mounts, and Escape while focus is inside the bar goes back to the report. Escape is handled on the bar's own element rather than a document listener, so it cannot swallow an Escape meant for something else on the page. Covered by two unit tests and a focus assertion in the e2e spec.

File: src/components/app-sidebar.tsx [hunk 4]
```diff
@@ -234,7 +234,8 @@ export function AppSidebar() {
   return (
     <Sidebar
       collapsible="icon"
-      className="shadow-lg"
+      // Ends above the screenshot bar when one is up, like the main panel.
+      className="shadow-lg h-[calc(100svh-var(--layout-bottom-bar-height))]"
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/app-sidebar.tsx]: Fixed in e38c5ec1. The contextual panel in app-sidebar.tsx and the lists in ChatList.tsx and AppList.tsx now subtract --layout-bottom-bar-height from their 100vh - 112px height, so nothing is clipped while the bar is up. The sidebar's own height also uses vh now, matching .h-screenish.

File: src/components/ScreenshotCaptureBar.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      observer?.disconnect();
+      style.removeProperty(HEIGHT_VAR);
+    };
+  }, []);
+
+  return (
+    <div
+      ref={root}
+      role="region"
+      aria-label={t("home:report.captureBarLabel")}
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Declining here. The Turkish locale (#4555) has no report section at all, so the whole bug report form is already shown in English for Turkish users on main. Adding just these five keys would put a Turkish bar over an English form. The right fix is a follow-up that translates the entire report section into Turkish and adds tr to rules/i18n.md, which still lists five locales.

[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Declining here, same reasoning as the Codex thread above: the Turkish locale has no report section at all, so the whole form is already English for Turkish users on main. Five bar keys alone would not help. A follow-up should translate the entire report section and add tr to rules/i18n.md.

**Comment by RyanGroch** [comment]
On the low-priority notes from the summary:

- Sidebar sub-list heights: fixed in e38c5ec1, the panel and both lists subtract the bar's height now.
- Mixed viewport units: fixed in e38c5ec1, the sidebar uses vh like .h-screenish.
- Unused common namespace: fixed in e38c5ec1.
- Silent half-second after Capture: leaving as is. Anything shown in that window would land in the screenshot, and the delay is what lets the native preview finish reflowing after the bar leaves the layout.

The three MEDIUM findings are answered in their own threads.

File: src/components/ScreenshotCaptureBar.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          {t("home:report.captureBarHeading")}
+        </span>{" "}
+        <span className="text-muted-foreground">
+          {t("home:report.captureBarHint")}
+        </span>
+      </p>
+      <div className="flex shrink-0 gap-1.5">
+        {/* The dialog this bar replaces took keyboard focus with it, and the
+            bar is last in tab order, so focus starts on the way forward. */}
+        <Button variant="default" size="sm" onClick={onCapture} autoFocus>
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Declining. The dialog's focus return goes through floating-ui's FloatingFocusManager, and its unmount cleanup only returns focus when the active element is inside the dialog or is document.body. When focus has already moved elsewhere it leaves it alone on purpose, per the comment there citing floating-ui issue 2607. The Capture button holds focus by the time the popup unmounts, so it is not taken back. Checked in the packaged app as well: Capture is still focused 1.5 seconds after the dialog closes, past the 200ms exit animation.

File: src/components/ScreenshotCaptureBar.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      aria-label={t("home:report.captureBarLabel")}
+      data-testid="screenshot-capture-bar"
+      // Only while focus is inside the bar, so it cannot swallow an Escape
+      // meant for something else on the page.
+      onKeyDown={(event) => {
+        if (event.key !== "Escape") return;
+        event.preventDefault();
+        onCancel();
+      }}
+      className="fixed inset-x-0 bottom-0 z-20 flex items-center gap-3 border-t-2 border-primary/70 bg-(--background-lightest) px-4 py-2 shadow-[0_-2px_8px_rgba(0,0,0,0.08)]"
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Fixed in 5d03297a. The Toaster now gets a bottom offset of sonner's own 24px plus --layout-bottom-bar-height, the same variable the privacy banner reads, so toasts stack above the bar while it is up and sit where they always did otherwise.

File: src/components/ScreenshotCaptureBar.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Declining, and this was verified by hand as suggested. The dialog's focus return runs through floating-ui's FloatingFocusManager, whose unmount cleanup only returns focus when the active element is inside the dialog or is document.body. When focus has already moved elsewhere it leaves it alone on purpose, per the comment there citing floating-ui issue 2607. In the packaged app, Capture is still the active element 1.5 seconds after the dialog closes, well past the 200ms exit animation.

File: src/components/ScreenshotCaptureBar.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Fixed in a09b4464 for the surfaces where a fix is possible: the bar is now z-[60], so it stays reachable over the fullscreen code view and the image lightbox.

Modal dialogs, alert dialogs and sheets remain unsupported on purpose. Base UI dialogs are modal by default and dismiss on any press outside the popup, so pressing Capture with the bar on top closes the dialog before the capture runs half a second later. That is not a regression: the old flow captured from inside the Help dialog, which is itself modal, so a second dialog was never in the picture. Capturing a modal would need a capture that is not a press, such as a keyboard shortcut the bar advertises, which is a follow-up rather than part of this change.

[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx]: Fixed in a09b4464 for the surfaces where a fix is possible: the bar is now z-[60], so it stays reachable over the fullscreen code view and the image lightbox.

Modal dialogs, alert dialogs and sheets remain unsupported on purpose. Base UI dialogs are modal by default and dismiss on any press outside the popup, so pressing Capture with the bar on top closes the dialog before the capture runs half a second later. That is not a regression: the old flow captured from inside the Help dialog, which is itself modal, so a second dialog was never in the picture. Capturing a modal would need a capture that is not a press, such as a keyboard shortcut the bar advertises, which is a follow-up rather than part of this change.

File: src/components/ScreenshotCaptureBar.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      // meant for something else on the page.
+      onKeyDown={(event) => {
+        if (event.key !== "Escape") return;
+        event.preventDefault();
+        onCancel();
+      }}
+      // Above the app's z-50 layer, so the bar stays reachable over the
+      // fullscreen code view and the image lightbox. Modal dialogs still
+      // close on the press, since it lands outside them.
+      className="fixed inset-x-0 bottom-0 z-[60] flex items-center gap-3 border-t-2 border-primary/70 bg-(--background-lightest) px-4 py-2 shadow-[0_-2px_8px_rgba(0,0,0,0.08)]"
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:72]: Declining for this PR. A countdown does not help: the press that starts it still lands outside the surface and dismisses it at once. The only real fix is a capture that is not a press, such as a keyboard shortcut the bar advertises, which is the follow-up already noted on the earlier z-index threads. On the double-click point, the bar sits above Base UI's internal backdrop, so a single press both dismisses the surface and starts the capture; no second click is needed. Leaving the hint copy as is: it already wraps at moderate widths, and this is an edge case of a flow that never supported capturing a second dialog.

File: e2e-tests/bug_report_screenshot.spec.ts [hunk 9]
```diff
@@ -123,11 +123,27 @@ test("report a bug with a chat session and a screenshot", async ({ po }) => {
...
+    // The dialog steps aside and leaves a bar in its place, so the reporter
+    // can go to wherever the bug is before capturing.
     await po.page.getByRole("button", { name: /Add a screenshot/ }).click();
+    const bar = po.page.getByTestId("screenshot-capture-bar");
+    await expect(bar).toBeVisible();
+    await expect(description).not.toBeVisible();
+    // Keyboard users land on the way forward, not on the page body.
+    await expect(
+      bar.getByRole("button", { name: "Capture screenshot" }),
+    ).toBeFocused();
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply e2e-tests/bug_report_screenshot.spec.ts:138]: Fixed in cc98c2a7. The spec now waits for the dialog count to reach zero before asserting focus, so the check runs after the dialog's own focus handling. In practice it already did, because the preceding check that the description is not visible only passes once the popup is unmounted, but now the ordering is explicit rather than incidental. The floating-ui side is covered on the other thread: its unmount cleanup leaves focus alone when it has already moved outside the dialog.

File: src/components/ScreenshotCaptureBar.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    return () => {
+      observer?.disconnect();
+      style.removeProperty(HEIGHT_VAR);
+    };
+  }, []);
+
+  return (
+    <div
+      ref={root}
+      role="region"
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:56]: Fixed in cc98c2a7. The instruction paragraph has an id and the Capture button points at it with aria-describedby, so when focus lands there a screen reader hears the heading and hint along with the button name. Covered by a unit test that follows the describedby id to the text.

File: src/components/ScreenshotCaptureBar.tsx [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+      // meant for something else on the page.
+      onKeyDown={(event) => {
+        if (event.key !== "Escape") return;
+        event.preventDefault();
+        onCancel();
+      }}
+      // Above the app's z-50 layer, so the bar stays reachable over the
+      // fullscreen code view and the image lightbox. Modal dialogs still
+      // close on the press, since it lands outside them.
+      className="fixed inset-x-0 bottom-0 z-[60] flex items-center gap-3 border-t-2 border-primary/70 bg-(--background-lightest) px-4 py-2 shadow-[0_-2px_8px_rgba(0,0,0,0.08)]"
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:72]: Declining. Capturing a modal is a deliberate scope decision for this PR, not an oversight, and the earlier threads on this file cover why: a press on the bar is an outside press, so any modal dismisses before the delayed capture, and the flow never supported a second dialog because the Help dialog is itself modal. The aria-hidden point holds only while a modal is open, which is that same unsupported case. Annotating the returned form would mean detecting an open modal at press time, which is fragile, and the reporter already sees the screenshot on the form before anything is sent, so nothing is sent unseen. A capture that is not a press, such as a keyboard shortcut the bar advertises, is the follow-up if this case matters.

**Review by wwwillchen — APPROVED** [review]
thanks! overall looks good to me. one thing that would be good is to remind users to paste the screenshot while they're creating the issue and right after they've created the issue. this is a pretty common issue as you can see: https://github.com/dyad-sh/dyad/issues?q=is%3Aissue%20%22reporter%20captured%20a%20screenshot%20in%20Dyad%22

this should be less critical once we automatically upload images, but for now a reminder would be good. thanks!

File: src/components/ScreenshotCaptureBar.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:72]: Declining. The bar is not inert while a modal is open. Base UI's FloatingFocusManager calls markOthers with the aria-hidden flag only and leaves the inert flag off, so outside elements are hidden from assistive tech but still receive pointer events. The press lands on the bar, counts as an outside press, and dismisses the modal, which is what the code comment describes. Capturing a modal is a deliberate scope decision for this PR, covered on the earlier threads on this file, so a test for the bar over a modal would only pin down a case that is declared unsupported. A capture that is not a press, such as a keyboard shortcut, remains the follow-up.

File: src/components/HelpDialog.tsx [hunk 12]
```diff
@@ -603,16 +610,38 @@ export function HelpDialog() {
...
+    setAwaitingCapture(false);
+    setHelpDialog({ open: true });
+  };
+
   const captureScreenshot = () => {
     if (isCapturing) return;
     const token = captureToken.current;
+    // The bar renders nothing while this is set, so it stays out of the
+    // picture.
     setIsCapturing(true);
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/HelpDialog.tsx:641]: Declining. Anything shown during that window lands in the screenshot, and the delay is what lets the native preview finish reflowing after the bar leaves the layout, so the gap is inherent to taking a clean capture. Re-showing the bar in a disabled state after the IPC is issued would put it back in the layout while the capture may still be running. On the Help button case: reopening the form mid-capture shows the add state for a moment, and the form then updates when the capture lands, because the capture token still matches. That needs a deliberate press inside a half-second window and resolves itself, so it does not justify an indicator.

File: src/app/layout.tsx [hunk 13]
```diff
@@ -210,12 +211,22 @@ function RootLayoutContent({ children }: { children: ReactNode }) {
                   richColors
                   expand
                   duration={settings?.isTestMode ? 500 : undefined}
+                  // Sonner's own 24px, plus the screenshot bar when one is up,
+                  // so toasts stack above the bar instead of covering its
+                  // buttons.
+                  offset={{
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/app/layout.tsx:217]: Fixed in cb3421e7, and you were right where I was wrong earlier. I had dismissed this on the 800px minimum window width, but the zoom setting goes to 150%, which makes that window a 533px CSS viewport, under sonner's 600px breakpoint. The Toaster now passes a mobileOffset of sonner's own 16px plus --layout-bottom-bar-height, matching the desktop offset.

File: src/components/ScreenshotCaptureBar.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:72]: Declining. The screenshot is not affected: the bar unmounts before the delayed capture runs, so nothing is hidden behind it in the image. What remains is that the bar covers the bottom strip of a fullscreen surface while the reporter is aiming, which is the trade-off taken when the bar moved to z-[60] so it stays reachable there at all. Shrinking the fullscreen code view and the lightbox means changing preview-panel surfaces this PR does not otherwise touch, for a strip that is back in view the moment Capture is pressed.

File: src/components/ScreenshotCaptureBar.tsx [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+      data-testid="screenshot-capture-bar"
+      // Only while focus is inside the bar, so it cannot swallow an Escape
+      // meant for something else on the page.
+      onKeyDown={(event) => {
+        if (event.key !== "Escape") return;
+        event.preventDefault();
+        onCancel();
+      }}
+      // Above the app's z-50 layer, so the bar stays reachable over the
+      // fullscreen code view and the image lightbox. Modal dialogs still
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:70]: Declining, same as the earlier threads on this file. Capturing a modal is a deliberate scope decision for this PR: a press on the bar is an outside press, the flow never supported a second dialog because the Help dialog is itself modal, and the reporter sees the resulting screenshot on the form before anything is sent. On the double-press concern, Base UI leaves the inert flag off for outside elements, so a single press both dismisses the modal and starts the capture. The bar copy stays as is because it already wraps at moderate widths. A capture that is not a press, such as a keyboard shortcut the bar advertises, is the follow-up if this case matters.

File: src/components/ScreenshotCaptureBar.tsx [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    <div
+      ref={root}
+      role="region"
+      aria-label={t("report.captureBarLabel")}
+      data-testid="screenshot-capture-bar"
+      // Only while focus is inside the bar, so it cannot swallow an Escape
+      // meant for something else on the page.
+      onKeyDown={(event) => {
+        if (event.key !== "Escape") return;
+        event.preventDefault();
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:65]: Fixed in 078c38e9. The bar's Escape handler now calls stopPropagation alongside preventDefault, so the key stops at the bar instead of reaching the fullscreen code view's window listener or the other global Escape listeners in the preview and chat layers. Covered by a unit test that registers a window keydown listener and checks it is not called, which I confirmed fails without the fix.

File: src/components/HelpDialog.tsx [hunk 16]
```diff
@@ -912,7 +949,7 @@ export function HelpDialog() {
               previewSrc={screenshotPreview}
               isCapturing={isCapturing}
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/HelpDialog.tsx:950]: Declining the removal, because the state is reachable. Opening Help from the sidebar while a capture is still in flight mounts the form with isCapturing true, since the capture token has not changed, and the form then shows the disabled Taking screenshot button and refuses a second capture. The existing unit test named keeps the button disabled when an older capture lands reaches exactly that state. Taking the fallback suggestion instead: as of 078c38e9 the prop carries a comment saying when the form can be up during a capture, so the next reader does not draw the same conclusion.

File: src/components/ScreenshotCaptureBar.tsx [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+      // close on the press, since it lands outside them.
+      className="fixed inset-x-0 bottom-0 z-[60] flex items-center gap-3 border-t-2 border-primary/70 bg-(--background-lightest) px-4 py-2 shadow-[0_-2px_8px_rgba(0,0,0,0.08)]"
+    >
+      <span className="flex h-8 w-8 shrink-0 items-center justify-center rounded-full bg-primary/15 text-primary">
+        <Camera className="h-4 w-4" />
+      </span>
+      <p id={instructionsId} className="min-w-0 flex-1 text-sm">
+        <span className="font-medium">{t("report.captureBarHeading")}</span>{" "}
+        <span className="text-muted-foreground">
+          {t("report.captureBarHint")}
```
[the remark this answers is not in the document]
Comment (RyanGroch) [reply src/components/ScreenshotCaptureBar.tsx:80]: Declining, as on the earlier threads on this file. Capturing a modal is a deliberate scope decision for this PR and is stated as unsupported in the PR description. On the mechanics: the dialogs here are Base UI, not Radix, and Base UI marks outside elements aria-hidden but leaves them clickable, so a single press on Capture both dismisses the modal and starts the capture. No second press is needed. The result is not silent either, since the reporter sees the screenshot on the form before anything is sent and can retake it. The hint copy stays as is because it already wraps at moderate widths. A capture that is not a press, such as a keyboard shortcut the bar advertises, is the follow-up if this case matters.
